File: 4284-smallest-stable-index-i/smallest-stable-index-i.java
class Solution {
    public int firstStableIndex(int[] nums, int k) {
        int n = nums.length;

        int [] right = new int [n];
        int M = Integer.MAX_VALUE;
        for(int i = n-1 ; i>=0; i--){
            M = Math.min(nums[i] , M);

            right[i] = M;
        }
        int []left = new int[n];
        int MAXI = Integer.MIN_VALUE;
        for(int i = 0 ;i<n ; i++){
            MAXI = Math.max(nums[i] , MAXI);
            left[i] = MAXI;
        }

        for(int i = 0 ; i < n ; i++ ){
            int diff = left[i] - right[i];

            if(diff <= k)
            return i;
        }

        return -1;
    }
}

File: 4284-smallest-stable-index-i/smallest-stable-index-i.cpp
class Solution {
public:
    int firstStableIndex(vector<int>& nums, int k) {
        int n = nums.size();
        int left = INT_MIN;
        vector <int> right(n);
        int M = INT_MAX;
        for(int i = n-1 ; i >= 0; i--){
            M = min(nums[i],M);
            right[i] = M;
        }

        for(int i = 0 ; i< n ; i++){
            left = max(left , nums[i]);

            if(left - right[i] <= k)
            return i;
        }

        return -1;
    }
};